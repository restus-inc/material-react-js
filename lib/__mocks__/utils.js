/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
