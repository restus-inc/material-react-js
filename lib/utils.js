/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
