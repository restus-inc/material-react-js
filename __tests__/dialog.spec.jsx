/* global jest, describe, afterEach, it, expect */
/* eslint object-curly-newline: ["error", { "minProperties": 5 }] */
import 'regenerator-runtime/runtime';
import React, { useState } from 'react';
import { render, cleanup, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { htmlOfRendering, getByOpenedDialog, getByDialogButton, getByDialogScrim } from './utils';

import Dialog from '../lib/dialog';

describe('Dialog component', () => {
  afterEach(async () => {
    await cleanup();
  });

  it('supports confirmation dialogs', () => {
    let buttons = [{ action: 'ok', label: 'OK', isDefault: true }];
    expect(htmlOfRendering((
      <Dialog title="foo" buttons={buttons} onClosed={() => {}}>
        <ul>
          <li>one</li>
          <li>two</li>
          <li>three</li>
        </ul>
      </Dialog>
    ))).resolves.toMatchSnapshot();

    buttons = [
      { action: 'cancel', label: 'Cancel', isDefault: true },
      { action: 'ok', label: 'OK' },
    ];
    expect(htmlOfRendering((
      <Dialog title="foo" className="bar" buttons={buttons} onClosing={() => {}}>
        <ol>
          <li>one</li>
          <li>two</li>
          <li>three</li>
        </ol>
      </Dialog>
    ))).resolves.toMatchSnapshot();
  });

  it('opens when isOpen property changes to true', async () => {
    let actionForClosing;
    const onOpening = jest.fn();
    const onOpened = jest.fn();
    const onClosing = jest.fn((event) => { actionForClosing = event.detail.action; });
    const onClosed = jest.fn((event) => { actionForClosing = event.detail.action; });
    const buttons = [
      { action: 'cancel', label: 'Cancel', isDefault: true },
      { action: 'ok', label: 'OK' },
    ];
    function DialogTester() {
      const [isOpen, setOpenStatus] = useState(false);
      return (
        <div>
          <button onClick={() => setOpenStatus(true)}>open</button>
          <Dialog title="foo" buttons={buttons} isOpen={isOpen}
                  onOpening={onOpening}
                  onOpened={onOpened}
                  onClosing={onClosing}
                  onClosed={(event) => { setOpenStatus(false); onClosed(event); }}>
            <ul>
              <li>one</li>
              <li>two</li>
              <li>three</li>
            </ul>
          </Dialog>
        </div>
      );
    }

    const { container, getByText } = render(<DialogTester/>);
    const openButton = getByText('open');
    const okButton = getByDialogButton(container, 'ok');
    const cancelButton = getByDialogButton(container, 'cancel');
    const scrim = getByDialogScrim(container);
    // An error occurs if the following function is mocked.
    okButton.getBoundingClientRect = jest.fn(() => ({ width: 100 }));

    expect(() => { getByOpenedDialog(); }).toThrow('Unable to find an opened dialog');
    expect(onOpening).not.toHaveBeenCalled();

    userEvent.click(openButton);
    expect(onOpening).toHaveBeenCalledTimes(1);
    expect(onOpened).not.toHaveBeenCalled();
    await waitFor(() => expect(onOpened).toHaveBeenCalledTimes(1));
    expect(onClosing).not.toHaveBeenCalled();
    expect(getByOpenedDialog()).toBeTruthy();

    userEvent.click(okButton);
    expect(onClosing).toHaveBeenCalledTimes(1);
    expect(actionForClosing).toBe('ok');
    expect(onClosed).not.toHaveBeenCalled();
    await waitFor(() => expect(onClosed).toHaveBeenCalledTimes(1));
    expect(actionForClosing).toBe('ok');
    expect(() => { getByOpenedDialog(); }).toThrow('Unable to find an opened dialog');

    expect(onOpening).toHaveBeenCalledTimes(1);
    userEvent.click(openButton);
    expect(onOpening).toHaveBeenCalledTimes(2);
    expect(onOpened).toHaveBeenCalledTimes(1);
    await waitFor(() => expect(onOpened).toHaveBeenCalledTimes(2));
    expect(getByOpenedDialog()).toBeTruthy();

    expect(onClosing).toHaveBeenCalledTimes(1);
    userEvent.click(cancelButton);
    expect(onClosing).toHaveBeenCalledTimes(2);
    expect(actionForClosing).toBe('cancel');
    expect(onClosed).toHaveBeenCalledTimes(1);
    await waitFor(() => expect(onClosed).toHaveBeenCalledTimes(2));
    expect(actionForClosing).toBe('cancel');
    expect(onOpening).toHaveBeenCalledTimes(2);
    expect(onOpened).toHaveBeenCalledTimes(2);
    expect(onClosing).toHaveBeenCalledTimes(2);
    expect(() => { getByOpenedDialog(); }).toThrow('Unable to find an opened dialog');

    userEvent.click(openButton);
    await waitFor(() => expect(onOpened).toHaveBeenCalledTimes(3));
    userEvent.click(scrim);
    expect(onClosing).toHaveBeenCalledTimes(3);
    expect(actionForClosing).toBe('close');
    expect(onClosed).toHaveBeenCalledTimes(2);
    await waitFor(() => expect(onClosed).toHaveBeenCalledTimes(3));
    expect(actionForClosing).toBe('close');
  });
});
