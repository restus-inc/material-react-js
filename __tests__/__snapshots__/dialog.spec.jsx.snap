// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component supports confirmation dialogs 1`] = `
"<div
  class=\\"mdc-dialog\\"
>
  <div
    class=\\"mdc-dialog__container\\"
  >
    <div
      aria-modal=\\"true\\"
      class=\\"mdc-dialog__surface\\"
      role=\\"alertdialog\\"
    >
      <h2
        class=\\"mdc-dialog__title\\"
      >
        foo
      </h2>
      <div
        class=\\"mdc-dialog__content\\"
      >
        <ul>
          <li>
            one
          </li>
          <li>
            two
          </li>
          <li>
            three
          </li>
        </ul>
      </div>
      <div
        class=\\"mdc-dialog__actions\\"
      >
        <button
          class=\\"mdc-button mdc-dialog__button\\"
          data-mdc-dialog-action=\\"ok\\"
          data-mdc-dialog-button-default=\\"\\"
          type=\\"button\\"
        >
          <div
            class=\\"mdc-button__ripple\\"
          />
          <span
            class=\\"mdc-button__label\\"
          >
            OK
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class=\\"mdc-dialog__scrim\\"
  />
</div>"
`;

exports[`Dialog component supports confirmation dialogs 2`] = `
"<div
  class=\\"mdc-dialog\\"
>
  <div
    class=\\"mdc-dialog__container\\"
  >
    <div
      aria-modal=\\"true\\"
      class=\\"mdc-dialog__surface bar\\"
      role=\\"alertdialog\\"
    >
      <h2
        class=\\"mdc-dialog__title\\"
      >
        foo
      </h2>
      <div
        class=\\"mdc-dialog__content\\"
      >
        <ol>
          <li>
            one
          </li>
          <li>
            two
          </li>
          <li>
            three
          </li>
        </ol>
      </div>
      <div
        class=\\"mdc-dialog__actions\\"
      >
        <button
          class=\\"mdc-button mdc-dialog__button\\"
          data-mdc-dialog-action=\\"cancel\\"
          data-mdc-dialog-button-default=\\"\\"
          type=\\"button\\"
        >
          <div
            class=\\"mdc-button__ripple\\"
          />
          <span
            class=\\"mdc-button__label\\"
          >
            Cancel
          </span>
        </button>
        <button
          class=\\"mdc-button mdc-dialog__button\\"
          data-mdc-dialog-action=\\"ok\\"
          type=\\"button\\"
        >
          <div
            class=\\"mdc-button__ripple\\"
          />
          <span
            class=\\"mdc-button__label\\"
          >
            OK
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class=\\"mdc-dialog__scrim\\"
  />
</div>"
`;
